qkeras/qconvolutional.py

```diff
@@ -509,6 +511,339 @@ def get_prunable_weights(self):
+    depthwise_kernel = array_ops.expand_dims(self.depthwise_kernel, 0)
+    pointwise_kernel = array_ops.expand_dims(self.pointwise_kernel, 0)
+    dilation_rate = (1,) + self.dilation_rate
+
+    if self.padding == 'causal':
+      op_padding = 'valid'
+    else:
+      op_padding = self.padding
+
+    if self.depthwise_quantizer:
+      quantized_depthwise_kernel = self.depthwise_quantizer_internal(
+          depthwise_kernel)
+    else:
+      quantized_depthwise_kernel = depthwise_kernel
+
+    if self.pointwise_quantizer:
+      quantized_pointwise_kernel = self.pointwise_quantizer_internal(
+          pointwise_kernel)
+    else:
+      quantized_pointwise_kernel = pointwise_kernel
+
+    outputs = tf.keras.backend.separable_conv2d(
+        inputs,
+        quantized_depthwise_kernel,
+        quantized_pointwise_kernel,
+        strides=self.strides * 2,
+        padding=op_padding,
+        dilation_rate=dilation_rate,
+        data_format=self.data_format)
+
+    if self.use_bias:
+      if self.bias_quantizer:
+        quantized_bias = self.bias_quantizer_internal(self.bias)
+      else:
+        quantized_bias = self.bias
+
+      outputs = tf.keras.backend.bias_add(
+          outputs,
+          quantized_bias,
+          data_format=self.data_format)
+
+    outputs = array_ops.squeeze(outputs, [spatial_start_dim])
+
+    if self.activation is not None:
+      return self.activation(outputs)
+    return outputs
+
+  def get_config(self):
+    config = {
+        "depthwise_quantizer":
+            constraints.serialize(self.depthwise_quantizer_internal),
+        "pointwise_quantizer":
+            constraints.serialize(self.pointwise_quantizer_internal),
+        "bias_quantizer":
+            constraints.serialize(self.bias_quantizer_internal)
+    }
+    base_config = super(QSeparableConv1D, self).get_config()
+    return dict(list(base_config.items()) + list(config.items()))
+
+  def get_quantizers(self):
+    return self.quantizers
+
+  def get_prunable_weights(self):
+    return [self.depthwise_kernel, self.pointwise_kernel]
+
+
+class QSeparableConv2D(SeparableConv2D, PrunableLayer):
+  """Depthwise separable 2D convolution."""
+
+  # most of these parameters follow the implementation of SeparableConv2D
+  # in Keras, with the exception of depthwise_quantizer, pointwise_quantizer
+  # and bias_quantizer.
+  #
+  # depthwise_quantizer: quantizer function/class for depthwise spatial kernel
+  # pointwise_quantizer: quantizer function/class for pointwise kernel
+  # bias_quantizer: quantizer function/class for bias
+  #
+  # we refer the reader to the documentation of SeparableConv2D in Keras for
+  # the other parameters.
+  #
+
+  def __init__(self,
+               filters,
+               kernel_size,
+               strides=(1, 1),
+               padding='valid',
+               data_format=None,
+               dilation_rate=(1, 1),
+               depth_multiplier=1,
+               activation=None,
```

Comment:
In the old version of `QSeparableConv2D` (now called `QMobileNetSeparableConv2D`), there were two activation parameters: `activation` (for pointwise activation), and `depthwise_activation` (for depthwise activation). However, in this implementation of `QSeparableConv2D` I believe there is only one `activation` parameter, which is used for pointwise activation. 

This means that the activation values after the depthwise conv but before the pointwise conv are kept in full-precision by Keras (see line 804). This may give the user a false impression that this layer is quantized, even though the activation values between the depthwise and pointwise conv are not quantized. For this reason, it may be better to support the use of a depthwise activation such that this quantization of activation values may occur. Perhaps the easiest way to do this may be to split the `tf.keras.backend.separable_conv2d` function into depthwise and pointwise convolutions with a depthwise activation function between them.

Reply:
`SeparableConv2D` in Keras doesn't apply activation between depthwise and pointwise steps, the [docs](https://keras.io/api/layers/convolution_layers/separable_convolution2d/) say the output is `activation(separableconv2d(inputs, kernel) + bias)`. Depthwise activation exists in `QMobileNetSeparableConv2D`, I kept that one around so users can have the activation if they want. I can add it to `QSeparableConv2D` (and remove the `QMobileNetSeparableConv2D`) but then it won't be 1:1 with Keras version and it would force downstream projects to implement `QSeparableConv2D` explicitly as a two step operation (but we already did implement it like that so it won't affect us atm)

Reply:
@vloncar Yes, you are right `SeparableConv2D` in Keras doesn't apply an activation function between the depthwise and pointwise steps - and we probably don't need to apply this activation function in QSeparableConv2D either. However, I think it is still important to quantize the activation values between depthwise and pointwise steps, otherwise QSeparableConv2D isn't completely quantized. An easy way to do this with QKeras would be to create a "linear" activation function that simply quantizes the activation values but does not apply any other non-linear activation functions. This is why I used the phrase "activation function" between the depthwise and pointwise steps, even though I mean a "quantization function" - sorry if that created any confusion.

Reply:
@vloncar,
I spoke with @zhuangh and I believe I better understand your motivation for the changes now. Since we have the `QMobileNetSeparableConv2D` layer, we don't necessarily need to add a `depthwise_activation` quantizer to the new `SeparableConv2D` class. However, I think it is worth putting a warning of sorts in the documentation that this SeparableConv2D does NOT quantize the activation values after the depthwise conv but before the pointwise conv. Beyond this, I think everything else is fantastic! Thank you!

Reply:
I can add the note/warning to the docs. I also need to fix the other issues before this can be merged. Sorry for the lack of activity, I've been busy, I'll complete this before the end of the month.

Reply:
I have finally rebased this and integrated the proposed changes. @danielemoro can you review? Thanks!